File: examples/csharp/Table/Agg/Handlers/CreateHandler.cs
using Google.Protobuf.WellKnownTypes;
using Angzarr.Client;
using Angzarr.Examples;

namespace Table.Agg.Handlers;

/// <summary>
/// Handler for CreateTable command.
/// </summary>
public static class CreateHandler
{
    public static TableCreated Handle(CreateTable cmd, TableState state)
    {
        // Guard
        if (state.Exists)
            throw CommandRejectedError.PreconditionFailed("Table already exists");

        // Validate
        if (string.IsNullOrEmpty(cmd.TableName))
            throw CommandRejectedError.InvalidArgument("table_name is required");
        if (cmd.SmallBlind <= 0)
            throw CommandRejectedError.InvalidArgument("small_blind must be positive");
        if (cmd.BigBlind <= 0)
            throw CommandRejectedError.InvalidArgument("big_blind must be positive");
        if (cmd.BigBlind < cmd.SmallBlind)
            throw CommandRejectedError.InvalidArgument("big_blind must be >= small_blind");
        if (cmd.MaxPlayers < 2 || cmd.MaxPlayers > 10)
            throw CommandRejectedError.InvalidArgument("max_players must be between 2 and 10");

        // Compute
        return new TableCreated
        {
            TableName = cmd.TableName,
            GameVariant = cmd.GameVariant,
            SmallBlind = cmd.SmallBlind,
            BigBlind = cmd.BigBlind,
            MinBuyIn = cmd.MinBuyIn != 0 ? cmd.MinBuyIn : cmd.BigBlind * 20,
            MaxBuyIn = cmd.MaxBuyIn != 0 ? cmd.MaxBuyIn : cmd.BigBlind * 100,
            MaxPlayers = cmd.MaxPlayers != 0 ? cmd.MaxPlayers : 9,
            ActionTimeoutSeconds = cmd.ActionTimeoutSeconds != 0 ? cmd.ActionTimeoutSeconds : 30,
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };
    }
}

File: examples/cpp/table/agg/handlers/create_handler.cpp
#include "create_handler.hpp"
#include "angzarr/errors.hpp"
#include <chrono>
#include <google/protobuf/util/time_util.h>

namespace table {
namespace handlers {

examples::TableCreated handle_create(
    const examples::CreateTable& cmd,
    const TableState& state) {

    // Guard
    if (state.exists()) {
        throw angzarr::CommandRejectedError::precondition_failed("Table already exists");
    }

    // Validate
    if (cmd.table_name().empty()) {
        throw angzarr::CommandRejectedError::invalid_argument("table_name is required");
    }
    if (cmd.small_blind() <= 0) {
        throw angzarr::CommandRejectedError::invalid_argument("small_blind must be positive");
    }
    if (cmd.big_blind() <= 0) {
        throw angzarr::CommandRejectedError::invalid_argument("big_blind must be positive");
    }
    if (cmd.big_blind() < cmd.small_blind()) {
        throw angzarr::CommandRejectedError::invalid_argument("big_blind must be >= small_blind");
    }
    if (cmd.max_players() < 2 || cmd.max_players() > 10) {
        throw angzarr::CommandRejectedError::invalid_argument("max_players must be between 2 and 10");
    }

    // Compute
    auto now = std::chrono::system_clock::now();
    auto timestamp = google::protobuf::util::TimeUtil::TimeTToTimestamp(
        std::chrono::system_clock::to_time_t(now));

    examples::TableCreated event;
    event.set_table_name(cmd.table_name());
    event.set_game_variant(cmd.game_variant());
    event.set_small_blind(cmd.small_blind());
    event.set_big_blind(cmd.big_blind());
    event.set_min_buy_in(cmd.min_buy_in() > 0 ? cmd.min_buy_in() : cmd.big_blind() * 20);
    event.set_max_buy_in(cmd.max_buy_in() > 0 ? cmd.max_buy_in() : cmd.big_blind() * 100);
    event.set_max_players(cmd.max_players() > 0 ? cmd.max_players() : 9);
    event.set_action_timeout_seconds(cmd.action_timeout_seconds() > 0 ? cmd.action_timeout_seconds() : 30);
    *event.mutable_created_at() = timestamp;

    return event;
}

} // namespace handlers
} // namespace table
